shared-module/adafruit_pixelbuf/PixelBuf.c

```diff
@@ -152,50 +152,56 @@ STATIC uint8_t _pixelbuf_get_as_uint8(mp_obj_t obj) {
         translate("can't convert %q to %q"), mp_obj_get_type_qstr(obj), MP_QSTR_int);
 }
 
-STATIC void _pixelbuf_parse_color(pixelbuf_pixelbuf_obj_t *self, mp_obj_t color, uint8_t *r, uint8_t *g, uint8_t *b, uint8_t *w) {
+STATIC color_u _pixelbuf_parse_color(pixelbuf_pixelbuf_obj_t *self, mp_obj_t color) {
+    color_u result;
     pixelbuf_byteorder_details_t *byteorder = &self->byteorder;
     // w is shared between white in NeoPixels and brightness in dotstars (so that DotStars can have
     // per-pixel brightness). Set the defaults here in case it isn't set below.
     if (byteorder->is_dotstar) {
-        *w = 255;
+        result.w = 255;
     } else {
-        *w = 0;
+        result.w = 0;
     }
 
     if (mp_obj_is_int(color) || mp_obj_is_float(color)) {
         mp_int_t value = mp_obj_is_int(color) ? mp_obj_get_int_truncated(color) : (mp_int_t)mp_obj_get_float(color);
-        *r = value >> 16 & 0xff;
-        *g = (value >> 8) & 0xff;
-        *b = value & 0xff;
+        result.r = value >> 16 & 0xff;
+        result.g = (value >> 8) & 0xff;
+        result.b = value & 0xff;
     } else {
         mp_obj_t *items;
         size_t len;
         mp_obj_get_array(color, &len, &items);
         mp_arg_validate_length_range(len, 3, 4, MP_QSTR_color);
 
-        *r = _pixelbuf_get_as_uint8(items[PIXEL_R]);
-        *g = _pixelbuf_get_as_uint8(items[PIXEL_G]);
-        *b = _pixelbuf_get_as_uint8(items[PIXEL_B]);
+        result.r = _pixelbuf_get_as_uint8(items[PIXEL_R]);
+        result.g = _pixelbuf_get_as_uint8(items[PIXEL_G]);
+        result.b = _pixelbuf_get_as_uint8(items[PIXEL_B]);
         if (len > 3) {
             if (mp_obj_is_float(items[PIXEL_W])) {
-                *w = 255 * mp_obj_get_float(items[PIXEL_W]);
+                result.w = 255 * mp_obj_get_float(items[PIXEL_W]);
             } else {
-                *w = mp_obj_get_int_truncated(items[PIXEL_W]);
+                result.w = mp_obj_get_int_truncated(items[PIXEL_W]);
             }
-            return;
+            return result;
         }
     }
     // Int colors can't set white directly so convert to white when all components are equal.
     // Also handles RGBW values assigned an RGB tuple.
-    if (!byteorder->is_dotstar && byteorder->bpp == 4 && byteorder->has_white && *r == *g && *r == *b) {
-        *w = *r;
-        *r = 0;
-        *g = 0;
-        *b = 0;
+    if (!byteorder->is_dotstar && byteorder->bpp == 4 && byteorder->has_white && result.r == result.g && result.r == result.b) {
+        result.w = result.r;
+        result.r = 0;
+        result.g = 0;
+        result.b = 0;
     }
+    return result;
 }
 
-STATIC void _pixelbuf_set_pixel_color(pixelbuf_pixelbuf_obj_t *self, size_t index, uint8_t r, uint8_t g, uint8_t b, uint8_t w) {
+STATIC void _pixelbuf_set_pixel_color(pixelbuf_pixelbuf_obj_t *self, size_t index, color_u rgbw) {
+    int r = rgbw.r;
```

Comment:
It's intentional, but I may be wrong.

On 32-bit arm micros, arithmetic on narrow types can actually be less efficient than arithmetic on wider types.  Canonical example [https://godbolt.org/z/Kr41o9sMd]:
```c
#include <stdint.h>

uint8_t add8(uint8_t a, uint8_t b) {
    return a+b;
}
uint32_t add32(uint32_t a, uint32_t b) {
    return a+b;
}
```
assembles to:
```
add8:
        adds    r0, r0, r1
        uxtb    r0, r0
        bx      lr
add32:
        adds    r0, r0, r1
        bx      lr
```
... because the uint8_t version must truncate after an arithmetic operation that might set high bits.

But reality begs to differ, or at least complicate.I double checked these results several times because they're surprising me:

On rp2040, which is cortex-m0plus with -O2 optimization, the "int" version is smaller by 8 bytes. On adafruit_proxlight_trinkey_m0 which is cortex-m0plus with -Os optimization, the "uint8_t" version is smaller by 24 bytes. So, the more size-efficient code depends on the compiler. As we are closer to the limit on builds where we've already switched to -Os, we should favor changes that give them the bigger size savings. The disassembly of the function in question is big enough that I didn't immediately puzzle out where the difference came in; possibly something about how multiplies or shifts are handled.

Reply:
perhaps but I'll leave it for now as "not worth it". It's important to us to size-optimize for the samd21 boards much more than anything else.